text/0000-single-formatter-for-all-contexts.md

```diff
@@ -0,0 +1,68 @@
+
+The backwards-incompatible changes come from the need to extract or inject context into new propagators. If a SaaS vendor or open source protocol previously had no need for on of the two propagators, but started to need it, they would need to notify all consumers that, as of this new version, the consumer must modify their code to add the additional formatter for the new context they require. This will probably come up often in SaaS offerings that previously only needed to propagate data in SpanContext, but then added their own fields that will then require custom DistributedContext formatters.
+
+## Explanation
+
+The new proposed API schema looks like:
+
+    formatters (namespace)
+        UnifiedContext.class
+        BinaryAPI.class
+        HTTPTextFormat.class
+
+With the following APIs:
+
+### UnifiedContext
+
+UnifiedContext will be a composite object containing all contexts that are present in the OpenTelemetry implementation. As of the authoring of this RFC, this object will contain a DistributedContext and SpanContext.
+
+### BinaryAPI
+
+The BinaryAPI will support two methods:
+
+    * fromBytes: convert raw bytes to a UnifiedContext or Null
+    * toBytes: convert a UnifiedContext to raw bytes
+
+### HTTPTextFormat
+
+The HTTPTextFormat will support two methods:
+
+    * extract: from a carrier object and getter, return a UnifiedContext
+    * inject: from a UnifiedContext, a carrier object, and a setter, return Null. populate the carrier object with values from the UnifiedContext.
+
+## Internal details
+
+This will require medium-sized refactoring on existing propagators and integrations.
+
+* integrations will have to switch to extracting the SpanContext and DistributedContext from the UnifiedContext.
+* all existing HTTPTextFormat and BinaryAPI APIs and implementations would have to refactored to match the unified HTTPTextFormat and BinaryAPIs.
+
+An example of the UnifiedContext at work (although not following the code organization above) exists at: 
+* PR: https://github.com/open-telemetry/opentelemetry-python/pull/89/files
+* file: https://github.com/open-telemetry/opentelemetry-python/blob/f1f8bb884fdcfe4ac257225e1636e1960495a51c/opentelemetry-api/src/opentelemetry/context/unified_context.py
+
+
+## Trade-offs and mitigations
+
+One drawback is this makes error cases a bit more ambiguous. For example, what happens if a format can extract a span properly, but not a distributed context? or visa versa?
```

Comment:
Another concern is how propagators for different wire formats can be composed. For example, I may want a service to support both Zipkin B3 format and W3C TraceContext format. In addition, Jaeger codecs support dedicated headers like `jaeger-debug-id` that can be set manually to force sampling of individual requests. There needs to be a way to compose different codecs.

Reply:
Yes! This is indeed an issue. 

I have an open issue in the spec about it: https://github.com/open-telemetry/opentelemetry-specification/issues/210. Happy to put this into an RFC as well, but there seems to be a line between what type of API changes can be proposed as an issue vs an RFC

Reply:
as a side note, I think issue is then you have a problem but not yet a solution and want to discuss it, whereas RFC is when you have both a problem (which may have been agreed upon in an issue) AND a proposed solution.

Reply:
Sounds good. I can move that to an RFC form.